src/cascadia/WinRTUtils/inc/SafeDispatcherTimer.h

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#pragma once
+
+// Par for the course, the XAML timer class is "self-referential". Releasing all references
+// to an instance will not stop the timer. Only calling Stop() explicitly will achieve that.
+struct SafeDispatcherTimer
+{
+    SafeDispatcherTimer() = default;
+    SafeDispatcherTimer(SafeDispatcherTimer const&) = delete;
+    SafeDispatcherTimer& operator=(SafeDispatcherTimer const&) = delete;
+    SafeDispatcherTimer(SafeDispatcherTimer&&) = delete;
+    SafeDispatcherTimer& operator=(SafeDispatcherTimer&&) = delete;
+
+    ~SafeDispatcherTimer()
+    {
+        Destroy();
+    }
+
+    explicit operator bool() const noexcept
+    {
+        return _timer != nullptr;
+    }
+
+    winrt::Windows::Foundation::TimeSpan Interval()
+    {
+        return _getTimer().Interval();
+    }
+
+    void Interval(winrt::Windows::Foundation::TimeSpan const& value)
+    {
+        _getTimer().Interval(value);
+    }
+
+    bool IsEnabled()
+    {
+        return _timer && _timer.IsEnabled();
+    }
+
+    void Tick(winrt::Windows::Foundation::EventHandler<winrt::Windows::Foundation::IInspectable> const& handler)
+    {
+        auto& timer = _getTimer();
+        if (_token)
+        {
+            timer.Tick(_token);
+        }
+        _token = timer.Tick(handler);
+    }
+
+    void Start()
+    {
+        _getTimer().Start();
+    }
+
+    void Stop() const
+    {
+        if (_timer)
+        {
+            _timer.Stop();
+        }
+    }
+
+    void Destroy()
+    {
+        if (!_timer)
+        {
+            return;
+        }
+
+        _timer.Stop();
+        if (_token)
+        {
+            _timer.Tick(_token);
+        }
+        _timer = nullptr;
+        _token = {};
+    }
+
+private:
+    ::winrt::Windows::UI::Xaml::DispatcherTimer& _getTimer()
+    {
+        if (!_timer)
+        {
+            _timer = ::winrt::Windows::UI::Xaml::DispatcherTimer{};
+        }
+        return _timer;
+    }
+
+    ::winrt::Windows::UI::Xaml::DispatcherTimer _timer{ nullptr };
+    winrt::event_token _token;
```

Comment:
somehow methinks there's an `auto_revoke` we could use here or something, but also, meh?

Reply:
Yeah I think `winrt::event_token` is better in this case. The `auto_revoke` classes are effectively just a `pair<weak_ref, event_token>` which, on destruction, turns it into a strong-ref and revokes it. Since this is such a tiny class, I think the extra work to get rid of the extra `weak_ref` is worth it. (It's just like 2 lines of extra code after all).